v0/destinations/refiner/utils.js

```diff
@@ -0,0 +1,210 @@
+} = require("../../../adapters/utils/networkUtils");
+const ErrorBuilder = require("../../util/error");
+const { isHttpStatusSuccess } = require("../../util/index");
+const { TRANSFORMER_METRIC } = require("../../util/constant");
+const { DESTINATION, CONFIG_CATEGORIES } = require("./config");
+const { getHashFromArray, getFieldValueFromMessage } = require("../../util");
+const { httpSend, prepareProxyRequest } = require("../../../adapters/network");
+
+/**
+ * Validation for userId and an email
+ * @param {*} message
+ * @returns
+ */
+const validatePayload = message => {
+  const userId = getFieldValueFromMessage(message, "userIdOnly");
+  const email = getFieldValueFromMessage(message, "email");
+  if (!userId && !email) {
+    throw new ErrorBuilder()
+      .setMessage("[Refiner] :: at least one param userId or email is required")
+      .setStatus(400)
+      .setStatTags({
+        destType: DESTINATION,
+        stage: TRANSFORMER_METRIC.TRANSFORMER_STAGE.TRANSFORM,
+        scope: TRANSFORMER_METRIC.MEASUREMENT_TYPE.TRANSFORMATION.SCOPE,
+        meta: TRANSFORMER_METRIC.MEASUREMENT_TYPE.TRANSFORMATION.META.BAD_EVENT
+      })
+      .build();
+  }
+};
+
+/**
+ * Replace the attributes name using webapp configuration mapping
+ * @param {*} attributesMap
+ * @param {*} destinationPayload
+ * @returns
+ */
+const replaceDestAttributes = (attributesMap, destinationPayload) => {
+  const payload = destinationPayload;
+  const keys = Object.keys(attributesMap);
+  keys.forEach(key => {
+    if (payload[key]) {
+      const value = payload[key];
+      payload[attributesMap[key]] = value;
+      delete payload[key];
+    }
+  });
+  return payload;
+};
+
+/**
+ * Returns account object
+ * ref : https://refiner.io/docs/api/#group-users
+ * @param {*} message
+ * @param {*} destination
+ * @returns
+ */
+const getAccountTraits = (message, destination) => {
+  const { accountAttributesMapping } = destination.Config;
+  const traits = getFieldValueFromMessage(message, "groupTraits");
+  const id = getFieldValueFromMessage(message, "groupId");
+  let groupUsersPayload = { ...traits, id };
+  const accountAttributesMap = getHashFromArray(
+    accountAttributesMapping,
+    "from",
+    "to",
+    false
+  );
+  groupUsersPayload = replaceDestAttributes(
+    accountAttributesMap,
+    groupUsersPayload
+  );
+  return groupUsersPayload;
+};
+
+/**
+ * Returns identify event payload
+ * ref : https://refiner.io/docs/api/#identify-user
+ * @param {*} message
+ * @param {*} destination
+ * @returns
+ */
+const identifyUserPayloadBuilder = (message, destination) => {
+  const { userAttributesMapping } = destination.Config;
+  const userId = getFieldValueFromMessage(message, "userIdOnly");
+  const email = getFieldValueFromMessage(message, "email");
+  const traits = getFieldValueFromMessage(message, "traits");
+  const contextTraits = get(message, "context.traits");
+  let payload = { ...traits, ...contextTraits, userId, email };
```

Comment:
we need to explicitly mention in doc as well as here that we are identifying existing users based on their remote_id, which we are mapping from userId and the email. So that if someone wants to update pre-existing users, they know what to send.